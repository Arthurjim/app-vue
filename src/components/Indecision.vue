<template>
    <img v-if="img" v-bind:src="img" alt="bg" />
    <div class="bg-dark"></div>
    <div class="indecision-container">
        <input
            type="text"
            v-model="question"
            placeholder="Hazme una pregunta"
        />
        <p>Recuerda terminar con un signo de interrogaci√≥n (?)</p>
        <div v-if="isValidateQuestion">
            <h2>{{question}}</h2>
            <h1>{{answer}}</h1>
        </div>
    </div>
</template>

<script>
export default {
    data() {
        return {
            question: "",
            answer:null,
            img:null,
            isValidateQuestion:false
        };
    },
    methods:{
       async getAnswer(){
         
           this.answer = "Pensando..."
            try {
            const {answer,image} =  await fetch('https://yesno.wtf/api')
            .then(res=> res.json())
                this.answer = answer === 'yes' ? 'Si!' : 'No!'
                this.img = image
            } catch (error) {
                console.log('Error')
                this.answer = 'Error en la api!'                
            }
        }
    },
    //Es para estar obsverando/pendiente a cambios
    watch:{
        question(value,oldValue){
            this.isValidateQuestion = false

            console.log({value})

            if(!value.includes('?'))return

            this.isValidateQuestion = true

            this.getAnswer()
            //TODO: Realizar la peticion http
        }
    }
};
</script>
<style scoped>
img,
.bg-dark {
    height: 100vh;
    left: 0px;
    max-height: 100%;
    max-width: 100%;
    position: fixed;
    top: 0px;
    width: 100vw;
}

.bg-dark {
    background-color: rgba(0, 0, 0, 0.4);
}

.indecision-container {
    position: relative;
    z-index: 99;
}

input {
    width: 250px;
    padding: 10px 15px;
    border-radius: 5px;
    border: none;
}
input:focus {
    outline: none;
}

p {
    color: rgb(214, 214, 214);
    font-size: 20px;
    margin-top: 0px;
}

h1,
h2 {
    color: white;
}

h2 {
    margin-top: 150px;
}
</style>
