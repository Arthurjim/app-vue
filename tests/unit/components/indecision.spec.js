import { shallowMount } from "@vue/test-utils";
import Indecision from "@/components/Indecision.vue";
describe("Pruebas en el componente Indecision", () => {
    let wrapper = shallowMount(Indecision);
    //Esta pendiente de cualquier metodo o atributo
    let getAnwerSpy = jest.spyOn(wrapper.vm, "getAnswer");

    //Creamos el mock fetch ya que no existe en node solo en el navegador
    global.fetch = jest.fn(() =>
        Promise.resolve({
            json: () =>
                Promise.resolve({
                    answer: "yes",
                    forced: false,
                    image: "https://yesno.wtf/assets/yes/2.gif",
                }),
        })
    );
    let clgSpy;
    beforeEach(() => {
        wrapper = shallowMount(Indecision);
        clgSpy = jest.spyOn(console, "log");
        getAnwerSpy = jest.spyOn(wrapper.vm, "getAnswer");

        //Limpiar todos los mocks
        jest.clearAllMocks();
    });
    test("Debe de estar igual al snaptshow", () => {
        expect(wrapper.html()).toMatchSnapshot();
    });
    test("Escribir en el input no debe de disparar nada console.log()", async () => {
        //Con el vm tenemos acceso a todas las instancias y metodos del componente
        const input = wrapper.find("input");
        await input.setValue("Hola mundo");

        expect(clgSpy).toHaveBeenCalledTimes(1);

        //Revisando que nunca se llame getAnswerSpy
        expect(getAnwerSpy).not.toHaveBeenCalled();
    });

    test("Se debe de disparar el getAnswer con el ?", async () => {
        const input = wrapper.find("input");
        await input.setValue("SerÃ© rico?");
        expect(getAnwerSpy).toHaveBeenCalledTimes(1);
    });

    test("Pruebas en getAnwer", async () => {
        await wrapper.vm.getAnswer();
        expect(wrapper.find("img")).toBeTruthy();
        expect(wrapper.vm.answer).toBe("Si!");
    });

    test('Pruebas en getAnwer -Fallo en API', async () => {
        //Hace el error en el fetch
        fetch.mockImplementationOnce(()=>Promise.reject('Api is down'))
        await wrapper.vm.getAnswer();

        expect(wrapper.find("img").exists()).toBeFalsy();
        expect(wrapper.vm.answer).toBe("Error en la api!");


        
     })
});
