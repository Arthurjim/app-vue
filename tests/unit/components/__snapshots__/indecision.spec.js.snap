// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente Indecision Debe de estar igual al snaptshow 1`] = `
<!--v-if-->
<div class="bg-dark"></div>
<div class="indecision-container"><input type="text" placeholder="Hazme una pregunta">
  <p>Recuerda terminar con un signo de interrogaci√≥n (?)</p>
  <!--v-if-->
</div>
`;
